File: Tayyibat/Services/TipsService.swift
import Foundation
import SwiftData

private struct TipBankFile: Codable {
    struct Entry: Codable { var category: String; var text: String }
    var version: String
    var tips: [Entry]
}

/// يدير بنك النصائح: التحميل، التخزين في SwiftData، والتدوير بدون تكرار.
enum TipsService {
    private static let recentWindow = 10

    /// يحمّل النصائح من JSON إلى SwiftData إذا لم تكن موجودة بعد.
    static func seedIfNeeded(context: ModelContext) {
        let count = (try? context.fetchCount(FetchDescriptor<NotificationTip>())) ?? 0
        guard count == 0 else { return }
        guard let url = Bundle.main.url(forResource: "tips_bank", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let bank = try? JSONDecoder().decode(TipBankFile.self, from: data) else {
            return
        }
        for entry in bank.tips {
            let category = TipCategory(rawValue: entry.category) ?? .general
            context.insert(NotificationTip(category: category, textAr: entry.text))
        }
        try? context.save()
    }

    /// نصيحة اليوم للرئيسية — ثابتة خلال اليوم وتتغيّر يومياً.
    static func tipOfTheDay(context: ModelContext) -> String {
        // ترتيب ثابت (بحسب المعرّف) كي لا تتغيّر نصيحة اليوم بين عمليات التشغيل.
        let pool = fetch(categories: [.morning, .general], context: context)
            .sorted { $0.id.uuidString < $1.id.uuidString }
        guard !pool.isEmpty else { return "تذكّر: كُل عند الجوع الحقيقي، واختر من الطيبات." }
        let dayIndex = Calendar.current.ordinality(of: .day, in: .year, for: .now) ?? 0
        return pool[dayIndex % pool.count].textAr
    }

    /// يختار نصيحة من فئة معيّنة متجنباً آخر النصائح التي ظهرت، ويعلّمها كمعروضة.
    @discardableResult
    static func nextTip(for category: TipCategory, context: ModelContext) -> String? {
        let pool = fetch(categories: [category], context: context)
        guard !pool.isEmpty else { return nil }

        let recentIDs = Set(recentlyShown(context: context).map(\.id))
        let candidates = pool.filter { !recentIDs.contains($0.id) }
        // الأقدم ظهوراً أولاً (nil يعني لم يظهر بعد).
        let chosen = (candidates.isEmpty ? pool : candidates)
            .sorted { ($0.lastShownAt ?? .distantPast) < ($1.lastShownAt ?? .distantPast) }
            .first

        chosen?.lastShownAt = .now
        try? context.save()
        return chosen?.textAr
    }

    // MARK: - Helpers

    private static func fetch(categories: [TipCategory], context: ModelContext) -> [NotificationTip] {
        let raws = Set(categories.map(\.rawValue))
        let all = (try? context.fetch(FetchDescriptor<NotificationTip>())) ?? []
        return all.filter { raws.contains($0.categoryRaw) }
    }

    private static func recentlyShown(context: ModelContext) -> [NotificationTip] {
        let all = (try? context.fetch(FetchDescriptor<NotificationTip>())) ?? []
        return all
            .filter { $0.lastShownAt != nil }
            .sorted { ($0.lastShownAt ?? .distantPast) > ($1.lastShownAt ?? .distantPast) }
            .prefix(recentWindow)
            .map { $0 }
    }
}

File: android/app/src/main/java/com/tayyibat/app/service/TipsService.kt
package com.tayyibat.app.service

import android.content.Context
import com.tayyibat.app.R
import com.tayyibat.app.data.db.TipDao
import com.tayyibat.app.data.model.NotificationTip
import com.tayyibat.app.data.model.TipCategory
import com.tayyibat.app.util.DateUtils
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.util.UUID

/** يدير بنك النصائح: التحميل، التخزين، والتدوير بدون تكرار. */
object TipsService {
    private const val RECENT_WINDOW = 10
    private val json = Json { ignoreUnknownKeys = true }

    @Serializable
    private data class TipBankFile(
        val version: String = "1.0",
        val tips: List<Entry> = emptyList(),
    ) {
        @Serializable
        data class Entry(val category: String, val text: String)
    }

    /** يحمّل النصائح من JSON إلى قاعدة البيانات إذا لم تكن موجودة بعد. */
    suspend fun seedIfNeeded(context: Context, dao: TipDao) {
        if (dao.count() > 0) return
        val raw = context.resources.openRawResource(R.raw.tips_bank)
            .bufferedReader().use { it.readText() }
        val bank = runCatching { json.decodeFromString(TipBankFile.serializer(), raw) }.getOrNull() ?: return
        val tips = bank.tips.map { entry ->
            NotificationTip(
                id = UUID.randomUUID().toString(),
                categoryRaw = (TipCategory.from(entry.category)).raw,
                textAr = entry.text,
            )
        }
        dao.insertAll(tips)
    }

    /** نصيحة اليوم للرئيسية — ثابتة خلال اليوم وتتغيّر يومياً. */
    suspend fun tipOfTheDay(dao: TipDao): String {
        val pool = dao.all()
            .filter { it.category == TipCategory.MORNING || it.category == TipCategory.GENERAL }
            .sortedBy { it.id }
        if (pool.isEmpty()) return "تذكّر: كُل عند الجوع الحقيقي، واختر من الطيبات."
        val dayIndex = DateUtils.dayOfYear()
        return pool[dayIndex % pool.size].textAr
    }

    /** يختار نصيحة من فئة معيّنة متجنباً آخر النصائح التي ظهرت، ويعلّمها كمعروضة. */
    suspend fun nextTip(category: TipCategory, dao: TipDao): String? {
        val all = dao.all()
        val pool = all.filter { it.category == category }
        if (pool.isEmpty()) return null

        val recentIds = all
            .filter { it.lastShownAt != null }
            .sortedByDescending { it.lastShownAt }
            .take(RECENT_WINDOW)
            .map { it.id }
            .toSet()

        val candidates = pool.filter { it.id !in recentIds }
        val chosen = (candidates.ifEmpty { pool })
            .minByOrNull { it.lastShownAt ?: Long.MIN_VALUE }
            ?: return null

        dao.update(chosen.copy(lastShownAt = DateUtils.now()))
        return chosen.textAr
    }
}
